buildSrc/subprojects/parser/src/main/kotlin/org/apache/calcite/buildtools/parser/DialectGenerateTask.kt

```diff
@@ -0,0 +1,316 @@
+        }
+    }
+
+    /**
+     * Parses a function of the form:
+     *
+     * <return_type> <name>(<args>) :
+     * {
+     *     <properties>
+     * }
+     * {
+     *     <body>
+     * }
+     *
+     * @param tokens The tokens starting from the function declaration and
+     *               ending at EOF
+     * @param functionMap The map to which the extracted functions will be added to
+     * @param charIndex The character index of the entire text of the file at which
+     *                  the parsing is commencing at
+     * @param declarationEnd The char index (of entire file text) at which the l
+     *                       function declaration ends
+     * @param functionName The name of the function being processed
+     */
+    private fun processFunction(
+        tokens: Queue<String>,
+        functionMap: MutableMap<String, String>,
+        charIndex: Int,
+        declarationEnd: Int,
+        functionName: String
+    ): Int {
+        var updatedCharIndex = charIndex
+        val functionBuilder = StringBuilder()
+        // Process the declaration:
+        while (updatedCharIndex < declarationEnd && tokens.isNotEmpty()) {
+            val token = tokens.poll()
+            functionBuilder.append(token)
+            updatedCharIndex += token.length
+        }
+        // Called twice as there are two curly blocks.
+        updatedCharIndex = processCurlyBlock(functionBuilder, tokens, updatedCharIndex)
+        updatedCharIndex = processCurlyBlock(functionBuilder, tokens, updatedCharIndex)
+
+        functionMap.put(functionName, functionBuilder.toString())
+        return updatedCharIndex
+    }
+
+    /**
+     * Parses a block of text surrounded by curly braces. The function keeps track
+     * of the number of curly braces that have yet to be closed. Once this counter
+     * reaches 0 the block has been fully parsed. The function also ensures that
+     * curly braces within single-line comments, multi-line comments, and strings
+     * are not counted for this.
+     *
+     * @param functionBuilder The builder unto which the tokens get added to once parsed
+     * @param tokens The tokens to be parsed
+     * @param charIndex The character index of the entire text of the file at which
+     *                  the parsing is commencing at
+     */
+    private fun processCurlyBlock(
+        functionBuilder: StringBuilder,
+        tokens: Queue<String>,
+        charIndex: Int
+    ): Int {
+        var curlyCounter = 0
+        var insideString = false
+        var insideCharacter = false
+        var insideSingleComment = false
+        var insideMultiComment = false
+        var updatedCharIndex = charIndex
+        while (tokens.isNotEmpty()) {
+            val token = tokens.poll()
+            functionBuilder.append(token)
+            updatedCharIndex += token.length
+            if (token == "\n" || token == " ") {
+                if (token == "\n") {
```

Comment:
Nested if's are typically discouraged where it's reasonably possible to replace them with something else. Can we just split the "\n" and " " logic into two separate clauses to avoid this?

Reply:
That's how it was initially but I thought this was better haha, will change back.